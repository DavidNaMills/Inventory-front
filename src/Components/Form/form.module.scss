@import '../../styles/variables/sizing.module.scss';
@import '../../styles/variables/colors.module.scss';

@mixin form {
    -webkit-backdrop-filter: blur(5px);
    backdrop-filter: blur(5px);
    // background-color: rgba(242, 242, 242, 0.5);

    margin-bottom: $largeSpacing;
    margin-top: $mediumSpacing;
    z-index: 1;
}



.form{
    border-radius: 3px;
    padding: 15px;
    
    
    -webkit-box-shadow: $blur $shadow-blue;
    -moz-box-shadow: $blur $shadow-blue;
    box-shadow: $blur $shadow-blue;

    &__buttonSpacing{
        &_large{
            margin-top: $mediumSpacing;
        }
        &_medium{
            margin-top: $extraSmallSpacing;
        }
    }
}

.blur{
    &__normal{
        width: 40%;
        @include form;
    }
    
    &_mobile{
        width: 80%;
        @include form;
    }
}




// @import '../../styles/variables/sizing.module.scss';
// @import '../../styles/variables/colors.module.scss';

// @mixin form {
//     border-radius: 3px;
//     padding: 15px;
//     margin-bottom: $largeSpacing;
//     margin-top: $mediumSpacing;
    
//     -webkit-box-shadow: $blur $shadow-blue;
//     -moz-box-shadow: $blur $shadow-blue;
//     box-shadow: $blur $shadow-blue;


// }


// .form{
//     &__normal{
//         width: 40%;
//         @include form;
//     }
    
//     &_mobile{
//         width: 80%;
//         @include form;
//     }

//     &__buttonSpacing{
//         &_large{
//             margin-top: $mediumSpacing;
//         }
//         &_medium{
//             margin-top: $extraSmallSpacing;
//         }
//     }
// }

// .blur{
//     // width:100%;
//     -webkit-backdrop-filter: blur(10px);
//     backdrop-filter: blur(5px);
//     background-color: rgba(255, 255, 255, 0.5); 
// }