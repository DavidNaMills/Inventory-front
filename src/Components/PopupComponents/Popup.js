import React from 'react';
import classes from './Popup.module.scss';
import usePositionHook from '../../hooks/usePositionHook/usePositionHook';

const style = (pos) => ({
    position: 'absolute',
    top: `${pos}px`,
    left: '0px',
    width: '100%',
    height: '100vh',
    backgroundColor: 'rgba(232, 255, 253, 0.6)',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    zIndex: '9999'
})

const Popup = (props) => {
    const { children, isOpen = false, onClose = null } = props;
    const { posRef } = usePositionHook();
    
    return (
        isOpen ?

            <div style={style(posRef)}>
                <div>
                    {children}
                </div>
            </div>
            : null
    )
}

export default Popup;