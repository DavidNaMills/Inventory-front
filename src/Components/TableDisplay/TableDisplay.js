import React from 'react';
import classes from './tableDisplay.module.scss';
import { useLocale } from '../../Context/LocaleContext';
import { isMobileOnly } from 'react-device-detect';
import StaffTypeRow from './Row/StaffTypeRow';
import ProductTypeRow from './Row/ProductTypeRow/ProductTypeRow';
import ProductRow from './Row/ProductRow/ProductRow';
import Button from '../Button/Button';
import MoreHorizIcon from '@material-ui/icons/MoreHoriz';
import CreateIcon from '@material-ui/icons/Create';


export const TableDisplay = (props) => {
    const locale = useLocale();
    const { list, config, update = null, remove = null, formName, viewMore = null, isSelected=null } = props;

    const all = list.map((x, count) => {
        let temp = [];
        const tStyle = [classes.tableDisplay];

        if (isSelected === x._id) { tStyle.push(classes.tableDisplay__selected) }
        else if (count % 2) { tStyle.push(classes.tableDisplay__even) }
        else { tStyle.push(classes.tableDisplay__odd) }

        switch (formName) {
            case 'staff':
                temp.push(
                    <StaffTypeRow
                        key={count}
                        config={config}
                        data={x}
                        showInfo={isMobileOnly ? false : true}
                    />
                )
                break;
            case 'productTypes':
                temp.push(
                    <ProductTypeRow
                        key={count}
                        config={config}
                        data={x}
                    />
                )
                break;

            case 'locationTypes':
                temp.push(
                    <ProductTypeRow
                        key={count}
                        config={config}
                        data={x}
                    />
                )
                break;
            case 'products':
                temp.push(
                    <ProductRow
                        key={count}
                        config={config}
                        data={x}
                        withImage={isMobileOnly? false : true}
                        withInfo={isMobileOnly? false : true}
                    />
                )
                break;
            default:
                return <p key={count}>ERROR</p>
        }
        return (
            <div className={tStyle.join(' ')} key={count}>
                {temp}
                <div className={classes.tableDisplay__optionsStyle}>
                    {update && temp.length > 0 &&
                        <Button onClick={() => update(JSON.parse(JSON.stringify(x)))} noStyle Icon={CreateIcon} />
                    }
                    {remove && temp.length > 0 && <Button onClick={() => remove(x._id)} type='danger' label={locale.remove} />}
                    {viewMore && <Button onClick={() => viewMore(x)} type='base' noStyle Icon={MoreHorizIcon} />}
                </div>
            </div>
        )
    })

    return (
        <div>
            { 
                all.length > 0
                ? all
                : <div className={classes.tableDisplay__message}><p className={classes.tableDisplay__message_text}>Nothing to display</p></div>
            }
            
        </div>
    );
}

export default TableDisplay;