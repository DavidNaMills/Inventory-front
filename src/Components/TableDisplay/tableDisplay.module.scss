@import '../../styles/variables/colors.module.scss';

.tableDisplay{
    display: flex;
    justify-content: space-between;
    padding: 5px;

    &__odd{
        background-color: $odd-back;
        border-bottom: solid $odd-border 2px;
    }

    &__even{
        background-color: $even-back;
        border-bottom: solid $even-border 2px;
    }

    &__selected{
        background-color: $selected-back;
    }

    &__optionsStyle {
        display: flex;
        justify-content: space-between;
        align-items: center;
        width: 150px;

        // margin-left: 15px;
        // margin-right: 15px
    }

    &__message{
        width: 100%;
        display: flex;
        justify-content: center;

        &_text{
            font-size: 30px;
            color: $title-color;
        }
    }
}