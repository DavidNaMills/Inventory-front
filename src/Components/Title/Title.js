import React from 'react';
import classes from './title.module.scss';


const Title = (props) => {
    const { size = null, message = '', center = false } = props;
    let element = null;

    switch (size) {
        case 'h1':
            element = <h1 className={[classes.title_h1, center ? classes.title__center : null].join(' ')} >{message}</h1>
            break;
        case 'h2':
            element = <h2 className={[classes.title_h2, center ? classes.title__center : null].join(' ')}>{message}</h2>
            break;
        case 'h3':
            element = <h3 className={[classes.title_h3, center ? classes.title__center : null].join(' ')}>{message}</h3>
            break;
        case 'h4':
            element = <h4 className={[classes.title_h4, center ? classes.title__center : null].join(' ')}>{message}</h4>
            break;
        case 'h5':
            element = <h5 className={[classes.title_h5, center ? classes.title__center : null].join(' ')}>{message}</h5>
            break;
        case 'h6':
            element = <h6 className={[classes.title_h6, center ? classes.title__center : null].join(' ')}>{message}</h6>
            break;
        default:
            element = <h2 className={[classes.title_h2, center ? classes.title__center : null].join(' ')}>{message}</h2>
            break;
    };

    return (
        <div>
            {element}
        </div>
    )
}

export default React.memo(Title);