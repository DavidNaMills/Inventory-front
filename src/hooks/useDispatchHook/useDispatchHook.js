import { useDispatch } from 'react-redux';

import { userLogin, userLogout, changeCurrentLocation } from '../../store/userStore/userActions';
import { startInitTypes, addNewProductType, addNewLocation, removeLocationType } from '../../store/typesStore/typesStoreActions';
import {createMessage} from '../../store/messagingStore/messagingActions';

const useDispatchHook = () => {
    const dispatch = useDispatch();

    const userLoginDispatch = (userData) => dispatch(userLogin(userData));
    const userLogoutDispatch = () => dispatch(userLogout());
    const changeCurrentLocationDispatch = (id) => dispatch(changeCurrentLocation(id));

    const initTypesDispatch = () => dispatch(startInitTypes());
    const addNewProductDispatch = (newType) => dispatch(addNewProductType(newType));
    const addNewLocationDispatch = (newLocation) => dispatch(addNewLocation(newLocation));
    const removeLocationDispatch = (id) => dispatch(removeLocationType(id));

    const createMessageDispatch = (data) => dispatch(createMessage(data));

    return {
        userLoginDispatch,
        userLogoutDispatch,
        changeCurrentLocationDispatch,

        initTypesDispatch,
        addNewProductDispatch,
        addNewLocationDispatch,
        removeLocationDispatch,

        createMessageDispatch
    }
}

export default useDispatchHook;