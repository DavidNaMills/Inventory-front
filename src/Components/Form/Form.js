import React from 'react';
import classes from './form.module.scss';
import { isMobileOnly } from 'react-device-detect';
import { useLocale } from '../../Context/LocaleContext';
import FormFactory from './Factory/FormFactory';
import Button from '../Button/Button';
import Title from '../Title/Title';
import PublishIcon from '@material-ui/icons/Publish';
import ClearIcon from '@material-ui/icons/Clear';

import objToArray from '../../helpers/objToArray/objToArray';


const Form = (props) => {
    const locale = useLocale();
    const {
        displayFields,
        user,
        fullForm,
        controlForm,
        updating,
        withTitle = true,

        submitCb = null,
        clearCb = null
    } = props;
    // isMobileOnly ? classes.form_mobile : classes.form__normal}
    return (
        <div className={isMobileOnly ? classes.blur_mobile : classes.blur__normal}>
            <div className={classes.form}>
                {withTitle && <Title message={updating ? `${locale.updating} ${updating}` : locale.creating} size='h3' />}
                <form onSubmit={submitCb}>
                    {
                        objToArray(fullForm).map((x, y) =>
                            <FormFactory
                                key={y}
                                config={x}
                                role={user.role}
                                displayFields={displayFields}
                                onChange={controlForm}
                            />
                        )
                    }
                    <div className={classes.form__buttonSpacing_large}>
                        <Button config={{ type: 'submit' }} type="submit" label={updating ? locale.update : locale.submit} Icon={PublishIcon} full />
                    </div>
                </form>
                {clearCb && <div className={classes.form__buttonSpacing_medium}>
                    <Button onClick={clearCb} label={locale.clear} type='clear' Icon={ClearIcon} full />
                </div>}
            </div>
        </div>
    )
}

export default Form;