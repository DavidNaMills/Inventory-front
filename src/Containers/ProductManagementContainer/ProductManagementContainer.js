import React, { useEffect, useState } from 'react';
import classes from '../../styles/common/containerForms.module.scss';
import { useSelector } from 'react-redux';
import {useLocale} from '../../Context/LocaleContext';
import { productFormConfig } from '../../configs/formConfigs/productFormConfig';

import useForm from '../../hooks/useForm/useForm';
import useDispatchHook from '../../hooks/useDispatchHook/useDispatchHook';
import useTableDisplay from '../../hooks/useTableDisplay/useTableDisplay';
import useAxios from '../../hooks/useAxios/useAxios';
import axios from '../../Axios/axiosInstance';
import { URLS } from '../../Axios/urls';

import Form from '../../Components/Form/Form';
import TableDisplay from '../../Components/TableDisplay/TableDisplay';
import Title from '../../Components/Title/Title';

import usePopup from '../../hooks/usePopup/usePopup';
import Popup from '../../Components/PopupComponents/Popup';
import ProductPopup from '../../Components/PopupComponents/ProductPopup';

import objToArray from '../../helpers/objToArray/objToArray';
import buildFormData from '../../helpers/buildFormData/buildFormData';
import productFactoryHelper from '../../helpers/productFactoryHelper/productFactoryHelper';
import Spinner from '../../Components/Spinner/Spinner';

const FORM_NAME = 'products';

const ProductManagementContainer = () => {
    const locale = useLocale();
    const { types } = useSelector(state => state);
    const { staff } = useSelector(state => state);

    const [displayFields, setDisplayFields] = useState(
        staff.config.formName[FORM_NAME].create
            ? staff.config.formName[FORM_NAME].create
            : staff.config.formName[FORM_NAME].update
    );
    const [isUpdate, setIsUpdate] = useState(null);

    const temp = objToArray(types.products);
    const {
        fullForm,
        buildFormObject,
        controlForm,
        clearForm,
        populateValues,
        validateFull,
        validatePartial
    } = useForm({ formConfig: productFormConfig(locale.errors), state: { products: temp } });

    const {fetchState, makeCall } = useAxios(axios);

    const {
        init,
        addItem,
        updateItem,
        toDisplayList
    } = useTableDisplay();

    const { createMessageDispatch } = useDispatchHook();
    const { openPopup, closePopup, isOpen } = usePopup();

    useEffect(() => {
        initFetch();
    }, [staff.currentLoc]);


    /**
     * Callbacks
     */

    const initFetch = () => {
        makeCall({
            url: `${URLS.allProducts}/${staff.currentLoc}/0/0`,
            method: 'GET',
            callBack: initDisplayTableCallback,
            newId: 'fetchInit'
        });
    }

    const initDisplayTableCallback = (res) => {
        if (res.error) {

        } else {
            init(res.products)
        }
    }

    const addCallback = (res) => {
        if (res.error) {
            console.log('something went wrong');
            console.log(res.error);
        } else {
            createMessageDispatch({ message: `${res.item.name} ${locale.succ_added}`, messageType: 'succType' });
            addItem(res.item);
            clearForm();
            resetUpdate();
        }
    }


    const updateCallback = (res) => {
        if (res.error) {
            console.log('something went wrong');
            console.log(res.error);
        } else {
            createMessageDispatch({ message: `${res.item.name} ${locale.succ_update}`, messageType: 'succType' });
            updateItem(res.item, res.item._id);
            clearForm();
            resetUpdate();
        }
    }

    const clearCb = () => {
        clearForm();
        if (isUpdate) {
            resetUpdate();
        }
    }

    /**
     * Clicks
     */


    const resetUpdate = () => {
        setIsUpdate(null);
        setDisplayFields(staff.config.formName[FORM_NAME].create);
    }

    const submitForm = (e) => {
        e.preventDefault();
        const valid = isUpdate ? validatePartial() : validateFull();
        if (valid) {
            const all = isUpdate
                ? productFactoryHelper({ data: buildFormObject(true), types: types.products, isNew: false })
                : productFactoryHelper({ data: buildFormObject(), usrLoc: staff.currentLoc, types: types.products })

            const temp = buildFormData(all);
            const conf = isUpdate
                ? { url: `${URLS.product}/${isUpdate._id}`, method: 'PUT', callBack: updateCallback }
                : { url: URLS.product, method: 'POST', callBack: addCallback }

            makeCall({
                ...conf,
                data: temp,
                newId: 'submit'
            });
        }

    }

    const update = (obj) => {
        const t = JSON.parse(JSON.stringify(obj));
        setIsUpdate(t);
        setDisplayFields(staff.config.formName[FORM_NAME].update);
        populateValues(obj);
        window.scrollTo({top: 0, behavior: 'smooth'});
    }


    const viewMore = (obj) => {
        openPopup(obj);
    }

    return (
        <div>
            {(fetchState.isLoading && fetchState.id === 'submit') && <Spinner/>}
            <Title size={'h1'} message={locale.cont_product} center/>

            <div className={classes.container}>
                <Form
                    formConfig={fullForm}
                    displayFields={displayFields}
                    controlForm={controlForm}
                    title={locale[staff.config.formName[FORM_NAME].title]}
                    formName={FORM_NAME}
                    user={staff.config}
                    fullForm={fullForm}
                    submitCb={submitForm}
                    clearCb={clearCb}
                    updating={isUpdate ? isUpdate.name : null}
                />
            </div>

            {staff.config.formName[FORM_NAME].tableDisplay 
                &&
                (fetchState.isLoading && fetchState.id === 'fetchInit')
                ? <Spinner withBackground={false}/>
                : <div className={classes.tableSpacing}>
                    <TableDisplay
                        list={toDisplayList}
                        config={staff.config.formName[FORM_NAME].tableDisplay.toDisplay}
                        update={staff.config.formName[FORM_NAME].tableDisplay.update ? update : null}
                        viewMore={viewMore}
                        formName={FORM_NAME}
                        isSelected={isUpdate ? isUpdate._id : null}
                    />
                </div>
                
            }

            <Popup isOpen={isOpen.show}>
                <ProductPopup onClose={closePopup} item={isOpen.selected} fields={staff.config.formName[FORM_NAME].tableDisplay.toDisplay.popup} />
            </Popup>

        </div>
    )
}

export default ProductManagementContainer;