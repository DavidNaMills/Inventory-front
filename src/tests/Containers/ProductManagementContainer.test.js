import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

import '../testHelpers/context';
import './containerHelper';

import Form from '../../Components/Form/Form';
import TableDisplay from '../../Components/TableDisplay/TableDisplay';
import Popup from '../../Components/PopupComponents/Popup';
import Title from '../../Components/Title/Title';
import ProductManagementContainer from '../../Containers/ProductManagementContainer/ProductManagementContainer';

jest.mock('../../hooks/useDispatchHook/useDispatchHook');
import useDispatchHook from '../../hooks/useDispatchHook/useDispatchHook';

const mockCreateMessageDispatch = jest.fn();

describe('<ProductManagementContainer/> test suite', ()=>{
    let wrapper;

    beforeEach(() => {
        useDispatchHook.mockImplementation(() => ({
            userLogoutDispatch: mockCreateMessageDispatch
        }));
        wrapper = shallow(<ProductManagementContainer/>);
    });

    it('matches snapshot', ()=>{
        const comp = renderer.create(<ProductManagementContainer />);
        const app = comp.toJSON();
        expect(app).toMatchSnapshot();
    });
    
    it('exists', ()=>{
        expect(wrapper.exists).toBeTruthy();
    });
    
    it('renders a <Title/> component', ()=>{
        const title = wrapper.find(Title);
        expect(title.length).toBe(1);
        expect(title.props()).toHaveProperty('size', 'h1');
        expect(title.props()).toHaveProperty('message', 'Product Management');
        expect(title.props()).toHaveProperty('center', true);
    });
    
    it('renders a <Form/> component', ()=>{
        const form = wrapper.find(Form);
        expect(form.length).toBe(1);
        expect(form.props()).toHaveProperty('formName', 'products');
    });
    
    it('renders a <TableDisplay/> component', ()=>{
        const td = wrapper.find(TableDisplay);
        expect(td.length).toBe(1);
        expect(td.props()).toHaveProperty('formName', 'products');
        expect(td.props()).toHaveProperty('isSelected', null);

    });
    
    it('renders a <Popup/> component', ()=>{
        expect(wrapper.find(Popup).length).toBe(1);
    });
});