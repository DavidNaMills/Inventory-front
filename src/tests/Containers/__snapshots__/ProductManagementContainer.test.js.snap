// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductManagementContainer/> test suite matches snapshot 1`] = `
<div>
  <div>
    <h1
      className="title_h1 title__center"
    >
      Product Management
    </h1>
  </div>
  <div
    className="container"
  >
    <div
      className="blur__normal"
    >
      <div
        className="form"
      >
        <div>
          <h3
            className="title_h3 "
          >
            Create new record
          </h3>
        </div>
        <form
          onSubmit={[Function]}
        >
          <div
            className="form__buttonSpacing_large"
          >
            <button
              className="button button__submit button_full"
              disabled={false}
              style={Object {}}
              type="submit"
            >
              <p>
                Submit
              </p>
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M5 4v2h14V4H5zm0 10h4v6h6v-6h4l-7-7-7 7z"
                />
              </svg>
            </button>
          </div>
        </form>
        <div
          className="form__buttonSpacing_medium"
        >
          <button
            className="button button__clear button_full"
            disabled={false}
            onClick={[Function]}
            style={Object {}}
          >
            <p>
              Clear
            </p>
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="tableSpacing"
  >
    <div />
  </div>
</div>
`;
